### PR TITLE
stmtdiagnostics: remove the usage of gossip

### PR DESCRIPTION
**Pull request by yuzefovich**
This commit removes the usage of gossip in the stmt diagnostics feature
since it is really an optimization (i.e not required for the feature to
work) and is in a way of the architecture unification (i.e.
multi-tenancy effort).

The gossip was previously used to let all nodes in the cluster know
about the new stmt diagnostics request or about the cancellation of an
existing one (in fact, the gossip propagation of the latter was broken
anyway since we forgot to register the corresponding callback). We now
rely on the polling mechanism on each node to read from the system
table to populate the registry of current requests. This polling takes
place every `sql.stmt_diagnostics.poll_interval` seconds (10 by default).

Additionally, this commit changes the class of that setting from
TenantWritable to TenantReadOnly because
1. we don't charge the user for the polling
2. to prevent the abuse where a tenant could set the polling interval
to a very low value incurring no costs to themselves (we excluded the
polling from the tenant cost recently).

The follow-up work remains to bring the optimization of propagating new
requests quickly using a range feed on the system table.

Epic: CRDB-16702
Informs: #47893

Release note: None

### CODE REVIEW

**Review by mgartner — APPROVED** [review]
:lgtm: But I'd like someone with more experience with this code to sign off on this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83547#-:-N5jeK3kCWgQAuAdfR70:b-8ll3xp)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz)

**Comment by yuzefovich** [comment]
I think Radu is the most familiar with this code. Maybe @irfansharif could take a quick look too given that you looked at this code recently?

**Comment by kevin-v-ngo** [comment]
> We could, but we might want to bring @kevin-v-ngo to this discussion. The question is whether it's worth making the stmt diagnostics feature have a larger latency (of propagation of the diagnostics requests throughout the cluster) in order to simplify the architecture unification effort.

Do we have a ballpark estimate on the impact to latency? But yes, sounds reasonable to me to be incremental and surface the functionality for CockroachDB Serverless and track a separate improvement on improving its performance.

**Comment by yuzefovich** [comment]
> Do we have a ballpark estimate on the impact to latency? But yes, sounds reasonable to me to be incremental and surface the functionality for CockroachDB Serverless and track a separate improvement on improving its performance.

Yes, that latency should have an upper bound determined by `sql.stmt_diagnostics.poll_interval` cluster setting (10 seconds by default).

**Review by irfansharif — APPROVED** [review]
Code changes LGTM. I'll defer to youse about this now being up to 10s stale, and maybe wanting to power this through rangefeeds for lower latency bounds. It's worth noting that unless we do something like https://github.com/cockroachdb/cockroach/issues/70614 and make the closed timestamp target duration a property of the table (and not the global cluster setting it is today), the staleness is going to be 3s, which is presumably still slower than gossip.

**Comment by ajwerner** [comment]
>the staleness is going to be 3s, which is presumably still slower than gossip.

I'm not sure how relevant this is. The events due to writes show up on rangefeeds at quite low-latency. You don't need the timestamp to be resolved to just get notified.

**Comment by andy-kimball** [comment]
I had thought [this PR](https://github.com/cockroachdb/cockroach/pull/83005) already excludes statement diagnostics from being charged for storage reads/writes. Are you seeing something different?

Is there some reason why statement diagnostics are _required_ for customers? If we allowed customers to turn off statement diagnostics, what problems would that cause? If they do have a way to turn them off, then perhaps we _should_ be charging them for the reads/writes.

**Comment by yuzefovich** [comment]
> I had thought https://github.com/cockroachdb/cockroach/pull/83005 already excludes statement diagnostics from being charged for storage reads/writes. Are you seeing something different?

Indeed, looks like we're good - I didn't see the change to the stmt diagnostics in #83005 before you pointed it out.

> Is there some reason why statement diagnostics are required for customers? If we allowed customers to turn off statement diagnostics, what problems would that cause? If they do have a way to turn them off, then perhaps we should be charging them for the reads/writes.

Before this PR it _was_ possible to turn off the polling mechanism of the stmt diagnostics, but the feature would still work in the single-tenant setting; however, in multi-tenant setting any attempt to create a new stmt diagnostics request would error out due to unavailability of gossip.

With this change I'm proposing to disallow the turning off of the polling so that the stmt diagnostics always worked. However, now that we exempt the polling queries from the tenant cost control, it seems possible that the polling interval could be set to a very low value, like 1ms, becoming a vector of abuse in the serverless setup. Maybe we ought to have a reasonable lower bound, say on the order of 1s to prevent such a scenario? We could also make the setting TenantReadOnly.

I could also see an argument for keeping the ability to turn it off - if a customer knows that they don't have any stmt diagnostics requests, it seems safe to turn the polling off entirely. Then we could be charging the users for reads and writes as you're suggesting. But that seems like a product decision.

**Comment by yuzefovich** [comment]
> Are RUs consumed for the statement diagnostics polling mechanism? So users always have some RU cost even if they don't activate statement diagnostics.

If this polling mechanism is not disabled, then yes, RUs will always be consumed, even if users don't ever use the stmt diagnostics feature.

**Comment by irfansharif** [comment]
Would use of a `rangefeedcache.Watcher` obviate the need for periodic scans of the table state? That way stmtdiagnostics can always be left enabled without accruing RUs.

**Comment by yuzefovich** [comment]
> Would use of a `rangefeedcache.Watcher` obviate the need for periodic scans of the table state?

Yeah, I guess so (not very familiar with that machinery). Irfan, do you have an estimate for how long implementing that would take?

**Comment by irfansharif** [comment]
Probably a day or two if you just cargo-cult the settingswatcher, which I think is the most barebones user of the library:

https://github.com/cockroachdb/cockroach/blob/8a77fe5f3bdcaba2fc63172bf0b7fd8a8f99dedc/pkg/server/settingswatcher/settings_watcher.go#L11-L13

**Comment by andy-kimball** [comment]
It'd be quite valuable for Serverless to get rid of the polling mechanism and instead only trigger background activity when we know there's probably work to be done.

**Review by yuzefovich — COMMENTED** [review]
I brought back the ability to disable the polling mechanism but changed the cluster setting class to TenantReadOnly. I think we should merge this change as is, and I _might_ work on the range feed mechanism (I took a quick look today, and it doesn't seem trivial, at least to me).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83547#-:-N6GR0vUAAJdbde6sSXQ:b2u4swv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz)

**Comment by yuzefovich** [comment]
Alright, so here is what I'm proposing with this PR:
- remove the usage of gossip entirely (this helps with the unification effort and enables the feature on the multi-tenant setup)
- keep the ability to disable the polling
- tenants don't get charged for polling the system table in Serverless (which is already the case, but previously they wouldn't be able to use the feature at all - we were just wasting the effort of polling)
- make the polling interval TenantReadOnly. cc @knz it'd be good to get your sign-off on this one.

Fully addressing #47893 by implementing a cache using a range feed over the system table in order to remove the polling mechanism completely will be done separately (with no timeline yet), so it is left as a follow-up. cc @vy-ton @andy-kimball for awareness.

If there are no objections (and the CI is green), I'll merge this PR tomorrow.

**Comment by yuzefovich** [comment]
> in the explanatory text (in commit message / PR description) you can also mention

Expanded the comment.

> who's filing the issue for the followup work?

I think it is already covered by #47893 (which is owned by SQL o11y).

**Pull request merged**
